### DISCUSSION TITLE
Pull Payments permissions / security

### DISCUSSION THREAD

**Discussion by ndeet** [btcpayserver/btcpayserver#4295] [Technical Support]
Working on the refunds for WooCommerce I have some questions about permissions. We only have the `btcpay.store.canmanagepullpayments` permission which will allow anybody with that permission to not only create pull payments but also make it skip the approval with `autoApproveClaims`. Playing devils advocate here if the site with the api key gets hacked or an employe claims the refunds himself. 

The only way to prevent this at the moment is to have a watch-only wallet afaics.

I think it would make a lot of sense to have separate permissions for CRUD pull payments and a separate one for approving them. This way your store can have a hot wallet but you can have a separate role that does the actual approval. This would mean that the `autoApproveClaims` only works when the API key has both permissions.

I'm not sure how this is handled with LN is it a normal LN payment that also has this approval step or is it using something like lnurl-withdraw and gets processed immediately? Anyway, it would be good to have it work the same for both cases.

**Comment by oskar321** [comment]
What's about refund approval with 2FA by admin or permitted user (employee)? If you are already working on it, if it is possible and not  very challenging :)

**Reply by ndeet** [reply]
Good idea but a bit out of scope for this maybe. I would think of it as a general additional security feature for approving transactions not only on pull payments?

**Reply by oskar321** [reply]
My focus was on the refund approval only but you're right. As a general feature it would be great. As you mentioned, playing devils advocate, nobody knows who to trust and how long. Any additional security option is welcome.

**Comment by Kukks** [comment]
We can have sub permissions as we do elsewhere. Maybe a cancreatepullpayments and then also a nested cancreatemanualpullpayments. Keep in mind this risk will still exist with employees. An employee with access to the store backend most likely has access to the btcpay backend, who could just approve payouts. In this case, we can have a user with "Guest" access to a btcpay store, and use his API keys to woocommerce?

**Reply by NicolasDorier** [reply]
Guest users aren't able to manage pull payment

**Reply by ndeet** [reply]
Not sure if the other sub-permissions are needed this is more about the approval step which I think we should have a more fine grained permission. So that the website owner can be sure that hackers (or employees) can't drain their wallet when they get hold of the API key. Even if I run the website and BTCPay myself I would prefer to make sure that the API key only can CRUD pull payments but never auto-approve so there is always another manual step needed. Websites running lots of plugins tend to have more vulnerabilities so it makes imo sense to separate those two things of CRUD and approval.

**Reply by NicolasDorier** [reply]
Note approval is only locking the rate if I remember correctly, it isn't sending money. (Except if the payout processor is running)

**Reply by ndeet** [reply]
So when I set the auto-approval flag no tx happens by default because no payout processor is running by default. If that is the case the question is still if we want a separate permissions so that one can use payout processors in combination with the API. E.g. a merchant might add several refunds over a day, then approves all claims manually and want to have the payment processor do its thing. Or is that not a use case and if one does manual approval then a manual batch tx is easy to do manually?

**Reply by ndeet** [reply]
Another thing to think about: In case of integrations we require the permissions on the setup routine so we would require the canmanagepullpayments permission upfront even if the user never uses that feature. They may be also not even aware of what this means but their API key would have needlessly the possibility to drain funds which would make them a good hacking target. So I would really prefer to have a either `cancreatepullpayments` permission like suggested by @kukks or an extra permission to make sure no pull payments are automatically paid out no matter what the payment processor config is.
